src/main/java/uk/gov/gchq/palisade/util/ResourceBuilder.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * Can produce any of the following output types:
+ * - {@link FileResource}
+ * - {@link DirectoryResource}
+ * - {@link SystemResource}
+ * Any parents automatically constructed will also be from this collection.
+ *
+ * If another method of creating a resource is required (i.e. directly using strings)
+ * there is no guarantee that this can correctly resolve parents. Instead use the
+ * methods provided by the appropriate resource impl.
+ */
+public class ResourceBuilder {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResourceBuilder.class);
+    private static final URI ROOT;
+
+    static {
+        File userDir = new File(System.getProperty("user.dir"));
+        URI root;
+        try {
+            root = userDir.getCanonicalFile().toURI();
+        } catch (IOException ex) {
+            root = userDir.getAbsoluteFile().toURI();
+        }
+        ROOT = root;
+    }
+
+    private ResourceBuilder() {
+        // empty private constructor
+    }
+
+    private enum Scheme {
+        file,
+        hdfs
+    }
+
+    public static boolean canCreate(final URI uri) {
+        try {
+            Scheme.valueOf(uri.getScheme()); // Or throw
+            return uri.isAbsolute();
+        } catch (IllegalArgumentException ex) {
+            return false;
+        }
+    }
+
+    // Create a leafResource from a uri, connectionDetail, type, serialisedFormat and attribute map
+    // Throw IllegalArgumentException if unsupported scheme
+    // Throw ClassCastException if uri did not point to a leaf resource
+    public static LeafResource create(final URI uri, final ConnectionDetail connectionDetail, final String type, final String serialisedFormat, final Map<String, String> attributes) {
+        return ((LeafResource) create(uri, attributes))
+                .connectionDetail(connectionDetail)
+                .type(type)
+                .serialisedFormat(serialisedFormat);
+    }
+
+    // Create a resource from a uri and attribute map
+    // Throw IllegalArgumentException if unsupported scheme
+    public static Resource create(final URI uri, final Map<String, String> attributes) {
+        // If passed relative paths, we can resolve them against the user.dir system property
+        URI absolute = uri.isAbsolute() ? uri : ROOT.resolve(uri);
+        // The hostname is all in the connectionDetail, so we never have a case of file://hostname/some/uri
+        // A lot of this is trying to normalize file:///some/uri (file://<no-hostname>/some/uri) to file:/some/uri
+        URI normal = UriBuilder.create(absolute)
+                    .withoutScheme()
+                    .withoutAuthority()
+                    .withoutPath()
+                    .withoutQuery()
+                    .withoutFragment();
+
+        // This should be assigning the attributes map to the returned object, once resources support attribute maps
+
+        switch (Scheme.valueOf(normal.getScheme())) {

```

Comment:
I think we need a check that if schema is not set then assume it is a file URI and act accordingly

Reply:
having looked into this I think we should leave it as is